File: src/mapBuilder/setTerminalValues.go
package main

import (
	"github.com/utkarsh-naman/3T/src/Constants"
	"github.com/utkarsh-naman/3T/src/model"
)

func main() {
	gamemap, _ := model.LoadGMap("bin/graph/unvalued/map0.ttt")
	setTerminal(gamemap)
}

func setTerminal(gamemap model.GMap) {
	terminalWinState := getTerminal(gamemap)

	for _, stateKey := range terminalWinState {
		props := gamemap[stateKey]
		props.Score = Constants.POSINF
		props.WinDepth = 0
		gamemap[stateKey] = props
	}

	//model.PrintGMap(gamemap)
	model.SaveGMap(gamemap, "bin/graph/valued/map1.ttt")
}

func getTerminal(gamemap model.GMap) (terminalWinStates []model.State) {
	var terminalWinMask model.State = 1 << 31
	var terminalTieMask model.State = 1 << 30

	for stateKey := range gamemap {
		if stateKey&terminalWinMask == 0 { // game over
			if stateKey&terminalTieMask == 0 { // game over & not a tie, therefore def a win
				terminalWinStates = append(terminalWinStates, stateKey)
			}
		}
	}
	return
}

File: Rust/src/mapbuilder/set_terminal_values.rs
use crate::constants::values::POSINF;
use crate::model::map::{GameMap};
use crate::model::state_model::State;

pub fn get_terminal_values(game_map: &GameMap) -> Vec<State> {
    let mut terminal_wins: Vec<State> = Vec::new();

    for key in game_map.keys() {
        if (*key>>31)&1 == 0{ // do not continue means terminal state
            if (*key>>30)&1 == 0 { // no tie so someone won
                terminal_wins.push(*key);
            }
        }
    }
    terminal_wins
}

pub fn set_terminal_values(mut map: GameMap) -> GameMap{
    let terminal_wins = get_terminal_values(&map);

    for state in terminal_wins{
        if let Some(state_props) = map.get_mut(&state) {
            state_props.score = POSINF;
            state_props.win_depth = 0;
        }
    }
    map
}